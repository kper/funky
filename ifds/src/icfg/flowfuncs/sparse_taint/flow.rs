use crate::icfg::state::State;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintNormalFlowFunction;

impl SparseTaintNormalFlowFunction {
    pub fn identity<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        fact: &Fact,
        pc: usize,
        defuse: &mut DefUseChain,
        edges: &mut Vec<Edge>,
    ) -> Result<()> {
        let after = defuse.get_next(ctx, function, &fact.belongs_to_var, pc)?;

        for to in after {
            edges.push(Edge::Normal {
                from: fact.apply(),
                to: to,
                curved: false,
            });
        }

        Ok(())
    }
}

impl SparseNormalFlowFunction for SparseTaintNormalFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        variable: &String,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Fact>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut facts = Vec::new();

        let instructions = &function.instructions;
        let instruction = instructions.get(pc);

        if instruction.is_none() {
            debug!("Instruction is none");
            return Ok(Vec::new());
        }

        let instruction = instruction.unwrap();

        debug!("Next instruction is {:?} for {}", instruction, variable);

        let mut nodes = defuse.demand(ctx, function, variable, pc)?;

        match instruction {
            Instruction::Unop(dest, ..) |
            Instruction::Phi(dest, ..) |
            Instruction::BinOp(dest, ..) | 
            Instruction::Assign(dest, ..) => {
                let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                nodes.extend(x.into_iter().map(|x| x.clone()));
            }
            _ => {}
        }
        let nodes = nodes
            .into_iter()
            .filter_map(|x| {
                if x.pc != x.next_pc {
                    Some(x.apply())
                } else {
                    Some(x)
                }
            })
            .collect::<Vec<_>>();

        facts.extend(nodes.into_iter());

        debug!("Out {:#?}", facts);

        Ok(facts)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::icfg::state::State;
    use crate::ir::ast::Program;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_binop() {
        /*
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %3 = %1 op %0
        */

        let func_name = "main".to_string();
        let function = AstFunction {
            name: func_name.clone(),
            definitions: vec![
                "%0".to_string(),
                "%1".to_string(),
                "%2".to_string(),
                "%3".to_string(),
            ],
            instructions: vec![
                Instruction::Const("%0".to_string(), 1.0),
                Instruction::Const("%1".to_string(), 1.0),
                Instruction::BinOp("%2".to_string(), "%0".to_string(), "%1".to_string()),
                Instruction::BinOp("%3".to_string(), "%0".to_string(), "%1".to_string()),
            ],
            ..Default::default()
        };

        let mut graph = Graph::default();
        let mut state = State::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            prog: &Program {
                functions: vec![function.clone()],
            },
            block_resolver: HashMap::default(),
        };

        let pc = 0;

        // fullfilling precondition of `chain.cache()`
        ctx.state.init_function(&function, pc).unwrap();

        let mut defuse = DefUseChain::default();

        let sparse = SparseTaintNormalFlowFunction;

        let facts = sparse
            .flow(&mut ctx, &function, 2, &"%0".to_string(), &mut defuse)
            .unwrap();

        debug!("facts {:#?}", facts);
        assert_eq!(4, facts.len());

        let cmp_facts: Vec<Fact> = vec![
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 3,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
        ];
        assert_eq!(cmp_facts, facts);
    }
}
