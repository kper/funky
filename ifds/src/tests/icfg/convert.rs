use crate::icfg::convert::ConvertSummary;
use crate::icfg::tikz::render_to;
use crate::solver::Request;
use insta::assert_snapshot;
use log::error;

use crate::grammar::*;

use crate::icfg::flowfuncs::taint::flow::TaintNormalFlowFunction;
use crate::icfg::flowfuncs::taint::initial::TaintInitialFlowFunction;

use std::fs::{OpenOptions, create_dir};
use std::io::Write;

/// Write the IR to a seperate file. This makes it possible
/// to run it in the UI.
fn write_ir(name: &str, ir: &str) {
    let _ = create_dir("src/tests/icfg/ir_code");
    let mut fs = OpenOptions::new()
        .write(true)
        .create(true)
        .open(format!("src/tests/icfg/ir_code/{}.ir", name))
        .unwrap();
    fs.write_all(&ir.as_bytes()).unwrap();
}

macro_rules! ir {
    ($name:expr, $req:expr, $ir:expr) => {
        let mut convert = ConvertSummary::new(TaintInitialFlowFunction, TaintNormalFlowFunction);

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let res = convert.visit(&prog, &$req);

        if let Err(err) = res {
            error!("ERROR: {}", err);
            err.chain()
                .skip(1)
                .for_each(|cause| error!("because: {}", cause));
            panic!("")
        }

        write_ir($name, $ir);

        let (graph, state) = res.unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("{}_dot", $name), output);
    };
}

#[test]
fn test_ir_const() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_const",
        req,
        "
         define test (result 0) (define %0) {
            %0 = 1
         };
    "
    );
}


#[test]
fn test_ir_simple_store() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_store",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            STORE %0 AT 0 + %0 ALIGN 2 32
         };
    "
    );
}

#[test]
fn test_ir_simple_load() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_load",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            %1 = LOAD %0 OFFSET 0 ALIGN 0
         };
    "
    );
}

#[test]
fn test_ir_double_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_double_const",
        req,
        "
         define test (result 0) (define %0 %1 %2){
            %0 = 1
            %1 = 1
            %2 = %0
            %2 = %1
         };
    "
    );
}

#[test]
fn test_ir_double_assign_with_params() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_double_const_with_params",
        req,
        "
         define test (param %0) (result 0) (define %0 %1 %2){
            %1 = 1
            %2 = %0
            %2 = %1
         };
    "
    );
}

#[test]
fn test_ir_chain_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_chain_assign",
        req,
        "
         define test (result 0) (define %0 %1 %2 %3){
            %0 = 1
            %1 = 1
            %2 = %0
            %3 = %2
         };
    "
    );
}

#[test]
fn test_ir_unop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_unop",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = op %0
            %1 = op %0   
        };"
    );
}

#[test]
fn test_ir_binop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_binop",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_binop_offset() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_binop_offset",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_phi() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_phi",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = phi %0 %1
        };"
    );
}

#[test]
fn test_ir_killing_op() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_killing_op",
        req,
        "define test (result 0) (define %0 %1 %2)  {
            %0 = 1
            %1 = 1
            KILL %0
            KILL %1
            %2 = 1
        };"
    );
}

#[test]
fn test_ir_block() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_block",
        req,
        "define test (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            GOTO 1
            BLOCK 1
            %1 = 2
        };"
    );
}

#[test]
fn test_ir_if_else() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_if_else",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = 2
            %2 = 3
            GOTO 3
            BLOCK 2
            %2 = 4
            GOTO 3
            BLOCK 3
            %0 = %2
        };
        "
    );
}

#[test]
fn test_ir_if() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_if",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 0
            %1 = 2
            %2 = 3
        };
        "
    );
}

#[test]
fn test_ir_loop() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_loop",
        req,
        "define main (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            %1 = 2
            GOTO 0 
        };
        "
    );
}

#[test]
fn test_ir_table() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_table",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            %1 = 2
            %2 = 3
            BLOCK 1
            %1 = 2
            %2 = 3
            BLOCK 2
            %2 = 4
            TABLE GOTO 0 1 2 ELSE GOTO 2
        };
        "
    );
}

#[test]
fn test_ir_functions() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_functions",
        req,
        "define test (result 0) (define %0) {
            %0 = 1
            CALL mytest(%0)
        };
        define mytest (param %0) (result 0) (define %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN;
        };"
    );
}

#[test]
fn test_ir_multiple_functions() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_multiple_functions",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0)  {
            RETURN %0;
        };"
    );
}

#[test]
fn test_ir_functions_rename_reg() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_functions_rename_regs",
        req,
        "define test (result 0) (define %0) {
            %0 = 1
            %0 <- CALL mytest(%0)
        };
        define mytest (param %5) (result 1) (define %5 %6)  {
            %6 = %5
            RETURN %6;
        };"
    );
}

#[test]
fn test_ir_return_values() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_values",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %1;
        };"
    );
}

#[test]
fn test_ir_return_passed_value() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_passed_value",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            RETURN %0;
        };"
    );
}

#[test]
fn test_ir_return_values2() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_values2",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 2
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %1 = 3
            RETURN %0;
        };"
    );
}

#[test]
fn test_ir_return_values3() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_values3",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 2
            %2 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            RETURN %0;
        };"
    );
}

#[test]
fn test_ir_overwrite_return_values() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_overwrite_return_values",
        req,
        "define test (result 0) (define %0) {
            %0 = 1
            %0 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %1;
        };"
    );
}

#[test]
fn test_ir_early_return() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_early_return",
        req,
        "define test (result 0) (define %0 %1 %2 %3) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 <- CALL mytestfoo(%0)
            %3 <- CALL mytestfoo(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %1;
        };
        define mytestfoo (param %0) (result 1) (define %0 %1) {
            %1 = 3
            RETURN %0;
        };
        "
    );
}

#[test]
fn test_ir_return_double() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %0 %1 <- CALL mytest(%0)
            %1 = 2
        };
        define mytest (param %0) (result 2) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %0 %1;
        };
        "
    );
}

#[test]
fn test_ir_return_branches() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_return_branches",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 5
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1 %2) {
            %1 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = 2
            %2 = 3
            RETURN %1;
            GOTO 3
            BLOCK 2
            %2 = 4
            GOTO 3
            BLOCK 3
            RETURN %0;
        };
        "
    );
}

#[test]
fn test_ir_self_loop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_self_loop",
        req,
        "define test (param %0) (result 1) (define %0 %1 %2) {
            %2 = 1
            %0 = 5
            %1 <- CALL test(%0)
            %0 = %1
            RETURN %0;
        };
        "
    );
}

#[test]
fn test_global_get_and_set() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_get_and_set",
        req,
        "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1) {
        %1 = %-1
        %-2 = %1
        };
    "
    );
}

#[test]
fn test_global_set() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_set",
        req,
        "
        define 0 (param %0) (result 0) (define %-1 %0 %1) {
            %0 = 1
            %-1 = %0
            %1 <- CALL 1()
        };

        define 1 (param) (result 1) (define %-1 %0) {
            %0 = %-1
            RETURN %0;
        };
    "
    );
}

#[test]
fn test_global_get_and_set_multiple_functions() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_get_and_set_multiple_functions",
        req,
        "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1 %2) {
        %1 = %-1
        %-2 = %1
        %2 = 1
        %0 <- CALL 1 (%2)
        %1 <- CALL 2 ()
        };

        define 1 (param %0) (result 1) (define %-2 %0) {
        %0 = %-2
        RETURN %0;
        };

        define 2 (result 1) (define %-2 %0) {
        %0 = 1
        %-2 = 1
        RETURN %0;
        };
    "
    );
}

#[test]
fn test_global_call() {
    let req = Request {
        variable: None,
        function: "1".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_global_call",
        req,
        "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1) {
        BLOCK 0
        %1 = %0
        %-1 = %1
        RETURN ;
        };
        define 1 (param %0) (result 0) (define %-2 %-1 %0 %1 %2) {
        BLOCK 1
        %1 = 1
        CALL 0(%1)
        %2 = %0
        %-2 = %2
        RETURN ;
        };
    "
    );
}

#[test]
fn test_global_writes() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_globals_writes",
        req,
        "
        define test (result 0) (define %-1 %0 %2) {
            %0 = 1
            %-1 = %0 
            %2 <- CALL mytest()
        };
        define mytest (param) (result 1) (define %-1 %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN %-1;
        };
    "
    );
}

#[test]
fn test_global_check_order() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_globals_check_order",
        req,
        "
        define test (result 0) (define %-2 %-1 %0 %2) {
            %0 = 1
            %-1 = %0 
            %2 <- CALL mytest()
        };
        define mytest (param) (result 1) (define %-2 %-1 %0 %1)  {
            %0 = 2   
            %1 = 3
            RETURN %-1;
        };
    "
    );
}

#[test]
fn test_memory_store() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_memory_store",
        req,
        "
        define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7 %8 %9) {
        BLOCK 0
        %1 = -12345
        STORE %1 AT 0 + %0 ALIGN 2 32
        %2 = 8
        %3 = -12345
        STORE %3 AT 0 + %2 ALIGN 3 64
        %5 = 8
        %6 = -12345
        STORE %6 AT 0 + %5 ALIGN 2 32
        %7 = 8
        %8 = -12345
        STORE %8 AT 0 + %7 ALIGN 3 64
        RETURN ;
        };
    "
    );
}

#[test]
fn test_memory_load() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_memory_load",
        req,
        "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE %1 AT 0 + %0 ALIGN 2 32
        %4 = 8
        %5 = LOAD %4 OFFSET 0 ALIGN 0
        %6 = 8
        %7 = LOAD %6 OFFSET 0 ALIGN 0
        KILL %7
        KILL %6
        RETURN ;
       }; 
    "
    );
}

#[test]
fn test_memory_load_different_functions() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_memory_load_different_functions",
        req,
        "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE %1 AT 0 + %0 ALIGN 2 32
        %2 <- CALL 1 ()
        RETURN ;
       }; 

       define 1 (result 1) (define %0 %1) {
        %1 = 8
        %0 = LOAD %1 OFFSET 0 ALIGN 0
        RETURN %0;
       };
    "
    );
}

#[test]
fn test_memory_load_different_functions2() {
    let req = Request {
        variable: None,
        function: "0".to_string(),
        pc: 2,
    };
    ir!(
        "test_ir_memory_load_different_functions2",
        req,
        "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE %1 AT 0 + %0 ALIGN 2 32
        %2 <- CALL 1 ()
        STORE %1 AT 1 + %0 ALIGN 2 32
        %3 <- CALL 1 ()
        RETURN ;
       }; 

       define 1 (result 1) (define %0 %1) {
        %1 = 8
        %0 = LOAD %1 OFFSET 0 ALIGN 0
        RETURN %0;
       };
    "
    );
}


