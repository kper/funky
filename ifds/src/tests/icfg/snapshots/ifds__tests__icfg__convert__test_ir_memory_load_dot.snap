---
source: src/tests/icfg/convert.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (0) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (1) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (2) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (3) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (4) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (5) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (6) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (7) at (0, 7) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (8) at (0, 8) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (9) at (0, 9) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (10) at (0, 10) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (11) at (0, 11) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (12) at (4, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (13) at (1, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (14) at (4, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (15) at (1, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (16) at (4, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%5$}] (17) at (3, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (18) at (1, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (19) at (4, 6) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%5$}] (20) at (3, 7) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (21) at (1, 7) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (22) at (4, 7) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%5$}] (23) at (3, 8) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%7$}] (24) at (2, 8) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (25) at (1, 8) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (26) at (4, 8) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%5$}] (27) at (3, 9) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (28) at (1, 9) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (29) at (4, 9) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%5$}] (30) at (3, 10) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (31) at (1, 10) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (32) at (4, 10) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%5$}] (33) at (3, 11) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $@0$}] (34) at (1, 11) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (35) at (4, 11) { };
\node[font=\tiny] (note_0) at (-1.5, 0.5) {Block(0)};
\node[font=\tiny] (note_1) at (-1.5, 1.5) {Const(0, 8.0)};
\node[font=\tiny] (note_2) at (-1.5, 2.5) {Const(1, -12345.0)};
\node[font=\tiny] (note_3) at (-1.5, 3.5) {Store(1, 0.0, 0)};
\node[font=\tiny] (note_4) at (-1.5, 4.5) {Const(4, 8.0)};
\node[font=\tiny] (note_5) at (-1.5, 5.5) {Load(5, 0.0, 4)};
\node[font=\tiny] (note_6) at (-1.5, 6.5) {Const(6, 8.0)};
\node[font=\tiny] (note_7) at (-1.5, 7.5) {Load(7, 0.0, 6)};
\node[font=\tiny] (note_8) at (-1.5, 8.5) {Kill(7)};
\node[font=\tiny] (note_9) at (-1.5, 9.5) {Kill(6)};
\node[font=\tiny] (note_10) at (-1.5, 10.5) {Return([])};
		\path[->, blue] (0) [loop right] edge  node { } (0);
		\path[->, blue] (0) [bend right] edge  node { } (1);
		\path[->, blue] (0) [bend right] edge  node { } (2);
		\path[->, blue] (0) [bend right] edge  node { } (3);
		\path[->, blue] (0) [bend right] edge  node { } (4);
		\path[->, blue] (0) [bend right] edge  node { } (5);
		\path[->, blue] (0) [bend right] edge  node { } (6);
		\path[->, blue] (0) [bend right] edge  node { } (7);
		\path[->, blue] (0) [bend right] edge  node { } (8);
		\path[->, blue] (0) [bend right] edge  node { } (9);
		\path[->, blue] (0) [bend right] edge  node { } (10);
		\path[->, blue] (0) [bend right] edge  node { } (11);
		\path[->, blue] (0) [bend right] edge  node { } (12);
		\path[->, blue] (0) [bend right] edge  node { } (13);
		\path[->, blue] (0) [bend right] edge  node { } (14);
		\path[->, blue] (0) [bend right] edge  node { } (15);
		\path[->, blue] (0) [bend right] edge  node { } (16);
		\path[->, blue] (0) [bend right] edge  node { } (17);
		\path[->, blue] (0) [bend right] edge  node { } (18);
		\path[->, blue] (0) [bend right] edge  node { } (19);
		\path[->, blue] (0) [bend right] edge  node { } (20);
		\path[->, blue] (0) [bend right] edge  node { } (21);
		\path[->, blue] (0) [bend right] edge  node { } (22);
		\path[->, blue] (0) [bend right] edge  node { } (23);
		\path[->, blue] (0) [bend right] edge  node { } (24);
		\path[->, blue] (0) [bend right] edge  node { } (25);
		\path[->, blue] (0) [bend right] edge  node { } (26);
		\path[->, blue] (0) [bend right] edge  node { } (27);
		\path[->, blue] (0) [bend right] edge  node { } (28);
		\path[->, blue] (0) [bend right] edge  node { } (29);
		\path[->, blue] (0) [bend right] edge  node { } (30);
		\path[->, blue] (0) [bend right] edge  node { } (31);
		\path[->, blue] (0) [bend right] edge  node { } (32);
		\path[->, blue] (0) [bend right] edge  node { } (33);
		\path[->, blue] (0) [bend right] edge  node { } (34);
		\path[->, blue] (0) [bend right] edge  node { } (35);


                \end{tikzpicture}
            \end{document}
